File: src/memory/builder/story_plan_preauthorization.py
"""Conditional Plan authority for implementable Ariad stories."""

from __future__ import annotations

from dataclasses import dataclass, replace
from pathlib import Path

from memory.builder.delivery_cursor import (
    BuilderDeliveryCursor,
    DeliveryCursorConflict,
    get_delivery_cursor,
    set_delivery_cursor,
)
from memory.builder.flow_unit import FLOW_UNIT_STORY_BY_STORY
from memory.builder.plan_preauthorization import (
    STORY_PLAN_CONTRACT,
    STORY_PLAN_REQUIRED_SECTIONS,
    PlanPreauthorizationMismatch,
    invalidate_plan_preauthorization,
    plan_preauthorization_mismatch_reason,
    unfilled_plan_sections_for,
)
from memory.builder.surface_protocol import wrap_ariad_surface
from memory.storage.store import Store


@dataclass(frozen=True)
class StoryPlanPreauthorizationReport:
    """Result of conditionally consuming one exact story Plan receipt."""

    cursor: BuilderDeliveryCursor
    status: str
    unfilled_sections: tuple[str, ...] = ()
    implementation_started: bool = False


def approve_story_plan_with_preauthorization(
    store: Store,
    *,
    journey: str,
    method: str,
    plan_artifact_path: Path | None,
) -> StoryPlanPreauthorizationReport:
    """Atomically consume exact story authority with Plan approval."""
    cursor = get_delivery_cursor(store, journey)
    if cursor is None:
        raise ValueError("delivery cursor is required before story Plan approval")
    if _is_consumed_story_approval(cursor):
        return StoryPlanPreauthorizationReport(cursor=cursor, status="already_approved")
    if cursor.active_item_level not in {"user_story", "technical_story"}:
        raise ValueError("story Plan preauthorization requires a User or Technical Story")
    if (
        cursor.active_checkpoint != "after_plan"
        or cursor.pending_confirmation != "navigator_approval"
    ):
        raise ValueError("story Plan approval requires a pending after_plan checkpoint")
    unfilled = unfilled_plan_sections_for(
        plan_artifact_path,
        required_sections=STORY_PLAN_REQUIRED_SECTIONS,
    )
    mismatch = plan_preauthorization_mismatch_reason(
        cursor,
        journey=journey,
        method=method,
        flow_unit=FLOW_UNIT_STORY_BY_STORY,
        child_work_items=(),
        plan_contract_version=STORY_PLAN_CONTRACT,
        unfilled_sections=unfilled,
    )
    if mismatch is not None:
        invalidate_plan_preauthorization(store, cursor, reason=mismatch)
        raise PlanPreauthorizationMismatch(mismatch)
    receipt = cursor.plan_preauthorization
    if receipt is None:  # pragma: no cover - mismatch guard establishes this
        raise PlanPreauthorizationMismatch("authorization_missing")
    try:
        updated = set_delivery_cursor(
            store,
            journey=journey,
            method=method,
            active_item=cursor.active_item,
            active_item_title=cursor.active_item_title,
            active_item_level=cursor.active_item_level,
            active_checkpoint=None,
            pending_confirmation=None,
            last_delivery_event="plan_approved",
            cadence_profile=cursor.cadence_profile,
            cadence_limits=cursor.cadence_limits,
            granularity_decision=cursor.granularity_decision,
            navigator_flow_unit=cursor.navigator_flow_unit,
            child_work_items=cursor.child_work_items,
            aggregate_checkpoint_status=cursor.aggregate_checkpoint_status,
            cursor_generation=cursor.cursor_generation,
            plan_preauthorization=replace(receipt, status="consumed", reason=None),
            expected_cursor=cursor,
            refresh_projection=False,
        )
    except DeliveryCursorConflict:
        current = get_delivery_cursor(store, journey)
        if current is not None and _is_consumed_story_approval(current):
            return StoryPlanPreauthorizationReport(cursor=current, status="already_approved")
        raise PlanPreauthorizationMismatch("cursor_changed") from None
    store.request_projection_refresh(journey)
    return StoryPlanPreauthorizationReport(
        cursor=updated,
        status="approved",
        unfilled_sections=unfilled,
        implementation_started=True,
    )


def cancel_story_plan_preauthorization(
    store: Store,
    *,
    journey: str,
    method: str,
) -> BuilderDeliveryCursor:
    """Cancel pending story authority while retaining ordinary Plan approval."""
    cursor = get_delivery_cursor(store, journey)
    if cursor is None:
        raise ValueError("delivery cursor is required before Plan preauthorization cancellation")
    if cursor.method != method:
        raise ValueError("active Builder method does not match cancellation method")
    if cursor.active_item_level not in {"user_story", "technical_story"}:
        raise ValueError("story Plan cancellation requires a User or Technical Story")
    receipt = cursor.plan_preauthorization
    if receipt is None or receipt.status != "pending":
        raise ValueError("pending Plan preauthorization is required before cancellation")
    return invalidate_plan_preauthorization(store, cursor, reason="navigator_cancelled")


def render_story_plan_preauthorization_recorded(cursor: BuilderDeliveryCursor) -> str:
    """Render story authority using bounded structural coordinates only."""
    receipt = cursor.plan_preauthorization
    if receipt is None:
        raise ValueError("Plan preauthorization receipt is required")
    body = "\n".join(
        [
            "Delivery",
            "",
            "╭────────────────────────────────────────────────────────╮",
            "│        🧭  PLAN PREAUTHORIZATION RECORDED              │",
            "│                                                        │",
            _card_text("Authorized story"),
            _card_text(receipt.active_item),
            "│                                                        │",
            _card_text("Story level"),
            _card_text(receipt.active_item_level),
            "│                                                        │",
            _card_text("Policy"),
            _card_text(receipt.policy),
            "│                                                        │",
            _card_text("Fixed stop"),
            _card_text(receipt.stop_boundary),
            "│                                                        │",
            _card_text("Boundary"),
            _card_text("Single-use authority remains pending until the"),
            _card_text("Driver completes the exact story Plan."),
            "╰────────────────────────────────────────────────────────╯",
        ]
    )
    return wrap_ariad_surface("plan_preauthorization_recorded", body + "\n")


def render_story_implementation_started(cursor: BuilderDeliveryCursor) -> str:
    """Render the single local implementation start after story approval."""
    body = "\n".join(
        [
            "Delivery",
            "",
            "╭────────────────────────────────────────────────────────╮",
            "│        🟧  IMPLEMENTATION STARTED                     │",
            "│                                                        │",
            _card_text("Active story"),
            _card_text(cursor.active_item or "none"),
            "│                                                        │",
            _card_text("Boundary"),
            _card_text("Local implementation may proceed under the exact"),
            _card_text("approved Plan and stops at Navigator Validation."),
            "╰────────────────────────────────────────────────────────╯",
        ]
    )
    return wrap_ariad_surface("implementation_started", body + "\n")


def render_story_preauthorization_already_consumed(cursor: BuilderDeliveryCursor) -> str:
    """Render an idempotent retry without another implementation start."""
    body = "\n".join(
        [
            "Delivery",
            "",
            "╭────────────────────────────────────────────────────────╮",
            "│        ✓  PLAN PREAUTHORIZATION ALREADY CONSUMED       │",
            "│                                                        │",
            _card_text("Active story"),
            _card_text(cursor.active_item or "none"),
            "│                                                        │",
            _card_text("Outcome"),
            _card_text("No approval or implementation start was repeated."),
            "╰────────────────────────────────────────────────────────╯",
        ]
    )
    return wrap_ariad_surface("plan_preauthorization_already_consumed", body + "\n")


def render_story_plan_preauthorization_mismatch(*, active_item: str | None, reason: str) -> str:
    """Render payload-free fallback to ordinary story Plan approval."""
    body = "\n".join(
        [
            "Delivery",
            "",
            "╭────────────────────────────────────────────────────────╮",
            "│        ✕  PLAN PREAUTHORIZATION MISMATCH               │",
            "│                                                        │",
            _card_text("Active story"),
            _card_text(active_item or "none"),
            "│                                                        │",
            _card_text("Reason"),
            _card_text(reason),
            "│                                                        │",
            _card_text("Fallback"),
            _card_text("Ordinary Navigator Plan approval remains required."),
            "╰────────────────────────────────────────────────────────╯",
        ]
    )
    return wrap_ariad_surface("plan_preauthorization_mismatch", body + "\n")


def _is_consumed_story_approval(cursor: BuilderDeliveryCursor) -> bool:
    receipt = cursor.plan_preauthorization
    return bool(
        receipt is not None
        and receipt.status == "consumed"
        and receipt.plan_contract_version == STORY_PLAN_CONTRACT
        and cursor.last_delivery_event == "plan_approved"
        and cursor.active_checkpoint is None
        and cursor.pending_confirmation is None
    )


def _card_text(text: str) -> str:
    value = text[:54]
    return f"│ {value:<54} │"

File: ts/src/builder/storyPlanPreauthorization.ts
// CV22.DS7.US8 plateau 3 — consuming story authority, once.
//
// Port of `src/memory/builder/story_plan_preauthorization.py`.
//
// The consume is a compare-and-swap against the observed cursor, and every branch
// around it is a safety property:
//
//   * an already-consumed receipt returns `already_approved` and starts NO second
//     implementation, so a retried command is idempotent rather than additive;
//   * a lost CAS re-reads the cursor and returns `already_approved` if another
//     writer got there first, and otherwise refuses with `cursor_changed`;
//   * a mismatch invalidates the receipt BEFORE raising, so a refused authority
//     cannot be retried into acceptance;
//   * the refusal path still leaves the Plan checkpoint pending, which is what makes
//     "fall back to ordinary Navigator approval" true rather than aspirational.

import type { WritableDatabase } from "#db/database.ts";
import { cardText } from "./card.ts";
import {
  type BuilderDeliveryCursor,
  type CursorWriteDeps,
  DeliveryCursorConflict,
  getDeliveryCursor,
  setDeliveryCursor,
  setTo,
} from "./deliveryCursor.ts";
import { FLOW_UNIT_STORY_BY_STORY } from "./flowUnit.ts";
import {
  invalidatePlanPreauthorization,
  PlanPreauthorizationMismatch,
  planPreauthorizationMismatchReason,
  STORY_PLAN_CONTRACT,
  STORY_PLAN_REQUIRED_SECTIONS,
  unfilledPlanSectionsFor,
} from "./planPreauthorization.ts";
import { wrapAriadSurface } from "./surfaceProtocol.ts";

const STORY_LEVELS = new Set(["user_story", "technical_story"]);

/** Python `StoryPlanPreauthorizationReport`. */
export interface StoryPlanPreauthorizationReport {
  readonly cursor: BuilderDeliveryCursor;
  readonly status: string;
  readonly unfilledSections: readonly string[];
  readonly implementationStarted: boolean;
}

/** Python `_is_consumed_story_approval`: all six conditions, or it is not consumed. */
function isConsumedStoryApproval(cursor: BuilderDeliveryCursor): boolean {
  const receipt = cursor.planPreauthorization;
  return (
    receipt !== null &&
    receipt.status === "consumed" &&
    receipt.planContractVersion === STORY_PLAN_CONTRACT &&
    cursor.lastDeliveryEvent === "plan_approved" &&
    cursor.activeCheckpoint === null &&
    cursor.pendingConfirmation === null
  );
}

export interface ApproveWithAuthorityOptions {
  readonly journey: string;
  readonly method: string;
  readonly planArtifactPath: string | null;
}

/** Python `approve_story_plan_with_preauthorization`. */
export function approveStoryPlanWithPreauthorization(
  db: WritableDatabase,
  options: ApproveWithAuthorityOptions,
  deps: CursorWriteDeps,
): StoryPlanPreauthorizationReport {
  const { journey, method } = options;
  const cursor = getDeliveryCursor(db, journey);
  if (cursor === null) throw new Error("delivery cursor is required before story Plan approval");
  if (isConsumedStoryApproval(cursor)) {
    return {
      cursor,
      status: "already_approved",
      unfilledSections: [],
      implementationStarted: false,
    };
  }
  if (cursor.activeItemLevel === null || !STORY_LEVELS.has(cursor.activeItemLevel)) {
    throw new Error("story Plan preauthorization requires a User or Technical Story");
  }
  if (
    cursor.activeCheckpoint !== "after_plan" ||
    cursor.pendingConfirmation !== "navigator_approval"
  ) {
    throw new Error("story Plan approval requires a pending after_plan checkpoint");
  }

  const unfilled = unfilledPlanSectionsFor(options.planArtifactPath, STORY_PLAN_REQUIRED_SECTIONS);
  const mismatch = planPreauthorizationMismatchReason(cursor, {
    journey,
    method,
    flowUnit: FLOW_UNIT_STORY_BY_STORY,
    childWorkItems: [],
    planContractVersion: STORY_PLAN_CONTRACT,
    unfilledSections: unfilled,
  });
  if (mismatch !== null) {
    invalidatePlanPreauthorization(db, cursor, mismatch, deps);
    throw new PlanPreauthorizationMismatch(mismatch);
  }

  const receipt = cursor.planPreauthorization;
  // Unreachable: the mismatch guard establishes a pending receipt. Kept because
  // Python keeps it, and because the type system cannot see that.
  if (receipt === null) throw new PlanPreauthorizationMismatch("authorization_missing");

  let updated: BuilderDeliveryCursor;
  try {
    updated = setDeliveryCursor(
      db,
      {
        journey,
        method,
        activeItem: cursor.activeItem,
        activeItemTitle: cursor.activeItemTitle,
        activeItemLevel: cursor.activeItemLevel,
        activeCheckpoint: null,
        pendingConfirmation: null,
        lastDeliveryEvent: "plan_approved",
        cadenceProfile: cursor.cadenceProfile,
        cadenceLimits: cursor.cadenceLimits,
        granularityDecision: cursor.granularityDecision,
        navigatorFlowUnit: cursor.navigatorFlowUnit,
        childWorkItems: cursor.childWorkItems,
        aggregateCheckpointStatus: cursor.aggregateCheckpointStatus,
        cursorGeneration: cursor.cursorGeneration,
        planPreauthorization: setTo({ ...receipt, status: "consumed", reason: null }),
        expectedCursor: cursor,
        refreshProjection: false,
      },
      deps,
    );
  } catch (error) {
    if (!(error instanceof DeliveryCursorConflict)) throw error;
    const current = getDeliveryCursor(db, journey);
    if (current !== null && isConsumedStoryApproval(current)) {
      return {
        cursor: current,
        status: "already_approved",
        unfilledSections: [],
        implementationStarted: false,
      };
    }
    throw new PlanPreauthorizationMismatch("cursor_changed");
  }

  deps.requestProjectionRefresh?.(journey);
  return {
    cursor: updated,
    status: "approved",
    unfilledSections: unfilled,
    implementationStarted: true,
  };
}

/** Python `cancel_story_plan_preauthorization`. */
export function cancelStoryPlanPreauthorization(
  db: WritableDatabase,
  options: { journey: string; method: string },
  deps: CursorWriteDeps,
): BuilderDeliveryCursor {
  const cursor = getDeliveryCursor(db, options.journey);
  if (cursor === null) {
    throw new Error("delivery cursor is required before Plan preauthorization cancellation");
  }
  if (cursor.method !== options.method) {
    throw new Error("active Builder method does not match cancellation method");
  }
  if (cursor.activeItemLevel === null || !STORY_LEVELS.has(cursor.activeItemLevel)) {
    throw new Error("story Plan cancellation requires a User or Technical Story");
  }
  const receipt = cursor.planPreauthorization;
  if (receipt === null || receipt.status !== "pending") {
    throw new Error("pending Plan preauthorization is required before cancellation");
  }
  return invalidatePlanPreauthorization(db, cursor, "navigator_cancelled", deps);
}

/** Python `render_story_plan_preauthorization_recorded`. */
export function renderStoryPlanPreauthorizationRecorded(cursor: BuilderDeliveryCursor): string {
  const receipt = cursor.planPreauthorization;
  if (receipt === null) throw new Error("Plan preauthorization receipt is required");
  const body = `${[
    "Delivery",
    "",
    "╭────────────────────────────────────────────────────────╮",
    "│        🧭  PLAN PREAUTHORIZATION RECORDED              │",
    "│                                                        │",
    cardText("Authorized story"),
    cardText(receipt.activeItem),
    "│                                                        │",
    cardText("Story level"),
    cardText(receipt.activeItemLevel),
    "│                                                        │",
    cardText("Policy"),
    cardText(receipt.policy),
    "│                                                        │",
    cardText("Fixed stop"),
    cardText(receipt.stopBoundary),
    "│                                                        │",
    cardText("Boundary"),
    cardText("Single-use authority remains pending until the"),
    cardText("Driver completes the exact story Plan."),
    "╰────────────────────────────────────────────────────────╯",
  ].join("\n")}\n`;
  return wrapAriadSurface("plan_preauthorization_recorded", body);
}

/** Python `render_story_implementation_started`. */
export function renderStoryImplementationStarted(cursor: BuilderDeliveryCursor): string {
  const body = `${[
    "Delivery",
    "",
    "╭────────────────────────────────────────────────────────╮",
    "│        🟧  IMPLEMENTATION STARTED                     │",
    "│                                                        │",
    cardText("Active story"),
    cardText(cursor.activeItem ?? "none"),
    "│                                                        │",
    cardText("Boundary"),
    cardText("Local implementation may proceed under the exact"),
    cardText("approved Plan and stops at Navigator Validation."),
    "╰────────────────────────────────────────────────────────╯",
  ].join("\n")}\n`;
  return wrapAriadSurface("implementation_started", body);
}

/** Python `render_story_preauthorization_already_consumed`. */
export function renderStoryPreauthorizationAlreadyConsumed(cursor: BuilderDeliveryCursor): string {
  const body = `${[
    "Delivery",
    "",
    "╭────────────────────────────────────────────────────────╮",
    "│        ✓  PLAN PREAUTHORIZATION ALREADY CONSUMED       │",
    "│                                                        │",
    cardText("Active story"),
    cardText(cursor.activeItem ?? "none"),
    "│                                                        │",
    cardText("Outcome"),
    cardText("No approval or implementation start was repeated."),
    "╰────────────────────────────────────────────────────────╯",
  ].join("\n")}\n`;
  return wrapAriadSurface("plan_preauthorization_already_consumed", body);
}

/** Python `render_story_plan_preauthorization_mismatch`: payload-free by construction. */
export function renderStoryPlanPreauthorizationMismatch(options: {
  activeItem: string | null;
  reason: string;
}): string {
  const body = `${[
    "Delivery",
    "",
    "╭────────────────────────────────────────────────────────╮",
    "│        ✕  PLAN PREAUTHORIZATION MISMATCH               │",
    "│                                                        │",
    cardText("Active story"),
    cardText(options.activeItem ?? "none"),
    "│                                                        │",
    cardText("Reason"),
    cardText(options.reason),
    "│                                                        │",
    cardText("Fallback"),
    cardText("Ordinary Navigator Plan approval remains required."),
    "╰────────────────────────────────────────────────────────╯",
  ].join("\n")}\n`;
  return wrapAriadSurface("plan_preauthorization_mismatch", body);
}
